pep-9999.rst

```diff
@@ -0,0 +1,874 @@
+already in use, and recommend that Publishing tools should gently nag users with
+informational warnings when they do not follow this PEP recommendations.
+
+This PEP is not about documenting license in code files, even though this is a
+surveyed topic in Appendix.
+
+
+Possible future PEPs
+====================
+
+It is the intention of the authors of this PEP to consider the submission of
+related but separate PEPs in the future such as:
+
+ - making the existing `License` and new `License-File` fields mandatory
+   including stricter enforcement in tools and Pypi publishing.
+
+ - restricting the upload of packages to the public Pypi index to the packaes
+   using only FOSS (Free and open source software) licenses.
+
+
+Motivation
+==========
+
+Software is licensed and providing accurate licensing information to Python
+packages users is an important matter.  Today, there are multiple places where
+license is documented in package metadata and there are limitations to what can
+be documented. This is often leading to confusion or a lack of clarity both for
+package authors and package users.
+
+Several package authors have expressed difficulty and/or frustrations with the
+possibilities to express licensing in package metadata. This also applies to
+Liux distribution packagers. This has triggered several license-related
+discussions and issues and in particular:
+
+- https://github.com/pypa/warehouse/issues/2996
+- https://github.com/pypa/interoperability-peps/issues/46
+- https://github.com/pypa/packaging-problems/issues/41
+- https://github.com/pypa/wheel/issues/138
+- https://github.com/pombredanne/spdx-pypi-pep/issues/1
+
+On average, Python packages tend to have more ambiguous or missing license
+information than other common application package formats (such as npm, Maven or
+Gem) as can be seen in the statistics [#cdstats]_ page of the ClearlyDefined
+[#cd]_ project that cover all packages from Pypi, Maven, npm and Rubygems.
+ClearlyDefined is an open source project to help improve clarity of other open
+source projecs that is incubating at the OSI (Open Source Initiative) [#osi]_.
+
+
+Rationale
+=========
+
+A mini survey of existing license metadata definitions in use in Python today
+and documented in several other system/distro and application package formats is
+provided as an informative Appendix in this PEP.
+
+There are a few takeaways from this survey:
+
+- Most package formats use a single `license` field.
+
+- Many modern package formats use some form of license expression syntax to
+  optionally combine more than one license identifiers together. SPDX and
+  SPDX-like syntaxes are the most popular in use.
+
+- SPDX license ids are becoming a de-facto way to reference common licenses
+  everywhere, whether or not a license expression syntax is used. But they often
+  need to be supplemented with extra license ids or conventions to accept extra
+  or generic licenses such as "Proprietary" or "Public domain" not tracked by
```

Comment:
- *Public domain* should be argued based on https://wiki.spdx.org/view/Legal_Team/Decisions/Dealing_with_Public_Domain_within_SPDX_Files

- I haven't ever seen good rationale why `Proprietary` needs to be there. It feels that in medium-size companies legal department tells tech department things, but nothing concrete.

- There's also https://github.com/spdx/spdx-spec/issues/49, i.e. you could say `NONE` to mean there are no license (and again, no lawyer have ever explained to me how `All rights reserved` differs from `license = NONE` in this technical context)`

- I don't see how `LicenseRef-PublicDomain` or `LicenseRef-Proprietrary` are any worse. They are however valid SPDX license expressions, so generic tools/libraries will understand them.

- Also, if there's some actual proprietary license, not "All rights reserved", then `LicenseRef-OurCompanyLicense` is valid license expression and more correct&descriptive.

Reply:
> LicenseRef-OurCompanyLicense is valid license expression

I spent a few minutes looking at the SPDX site and couldn't find any confirmation of this (not saying you're incorrect, but rather that the information is hard to find) and there's no way I'd have known to even look for an expression like this if all I knew was "I need to record that the package I'm publishing is licensed under our company license".

My concern here is that if it's too hard for people to find a reasonable thing to put in this metadata, they'll end up not bothering, and just supplying a `LICENSE.txt` file, or if the field is made mandatory we'll get endless support requests on the packaging tracker.

Reply:
Look at appendix iv in  https://spdx.org/spdx-specification-21-web-version

> Some examples:  
> LicenseRef-23
> LicenseRef-MIT-Style-1
> DocumentRef-spdx-tool-1.2:LicenseRef-MIT-Style-2

Unfortunaty, indeed https://spdx.org/ids-how doesn't mentioned `LicenseRef` and `DocumentRef` use cases directly. That's something to be raised with SPDX people.

Reply:
@pfmoore if company went through a cost of writing own proprietary license, it's not hard to figure out what SPDX expression should they use for it.

I repeat: SPDX is univeral standard, having Python-specific deviation won't really help anyone.

Reply:
@phadej FWIW, I happen to be one of the SPDX co-founders though I speak here exclusively with my Python hat on and not on behalf of SPDX. 

`LicenseRef-XXX` license ids are only valid within the context of a full SPDX document and not outside in a solo expression like we are considering here. So using  `LicenseRef-Proprietary` is no more valid than using `Proprietary` in this context ... much it is much simpler to write down, remember and specify. 

There are some ongoing discussions at SPDX to defines license ids "namespaces" to cope with this but this is not fully there yet. In the meantime, there is no many other ways than to use expressions with extra not-SPDX-listed ids (which is what `npm` and `Suse` and `ClearlyDefined` do for now)..

@pf_moore you wrote:
> My concern here is that if it's too hard for people to find a reasonable thing to put in this metadata, they'll end up not bothering, and just supplying a LICENSE.txt file, or if the field is made mandatory we'll get endless support requests on the packaging tracker.

exactly: which is why I am not making anything mandatory and I find that using a generic `Proprietary` for anything off SPDX is simpler.